HLTrigger/btau/src/HLTDisplacedtktktkFilter.cc

```diff
@@ -0,0 +1,210 @@
+  iEvent.getByToken(TrackToken_,trackcands);
+
+  // All HLT filters must create and fill an HLT filter object,
+  // recording any reconstructed physics objects satisfying (or not)
+  // this HLT filter, and place it in the Event.
+
+
+  // Ref to Candidate object to be recorded in filter object
+  reco::RecoChargedCandidateRef ref1;
+  reco::RecoChargedCandidateRef ref2;
+  reco::RecoChargedCandidateRef ref3;
+
+  if (saveTags()) filterproduct.addCollectionTag(TrackTag_);
+
+  bool triggered = false;
+
+  // loop over vertex collection
+  for(auto displacedVertex : displacedVertexColl){
+    // check if the vertex actually consists of exactly three track tracks, reject the event if not
+    if(displacedVertex.tracksSize() != 3){
+      edm::LogError("HLTDisplacedtktktkFilter") << "HLTDisplacedtktktkFilter: ERROR: the tktktk vertex must have exactly three tracks by definition. It now has n tracks = "<< displacedVertex.tracksSize() << std::endl;
+      return false;
+    } 
+
+    float normChi2 = displacedVertex.normalizedChi2();
+    if (normChi2 > maxNormalisedChi2_) continue;
+
+    double vtxProb = 0.0;
+    if( (displacedVertex.chi2()>=0.0) && (displacedVertex.ndof()>0) ) vtxProb = TMath::Prob(displacedVertex.chi2(), displacedVertex.ndof() );
+    if (vtxProb < minVtxProbability_) continue;
+
+    // get the three tracks from the vertex
+    auto trackIt =  displacedVertex.tracks_begin();
+    reco::TrackRef vertextkRef1 =  (*trackIt).castTo<reco::TrackRef>();
+    // the second one
+    trackIt++;
+    reco::TrackRef vertextkRef2 =  (*trackIt).castTo<reco::TrackRef>();
+    // the third one
+    trackIt++;
+    reco::TrackRef vertextkRef3 =  (*trackIt).castTo<reco::TrackRef>();
+
+    // first find these three tracks in the track collection
+    reco::RecoChargedCandidateCollection::const_iterator cand1;
+    reco::RecoChargedCandidateCollection::const_iterator cand2;	
+    reco::RecoChargedCandidateCollection::const_iterator cand3;	
+
+    int iFoundRefs = 0;
+    for (auto cand=trackcands->begin(); cand!=trackcands->end(); cand++) {
+      reco::TrackRef tkRef = cand->get<reco::TrackRef>();
+      if(tkRef == vertextkRef1) {cand1 = cand; iFoundRefs++;}
+      if(tkRef == vertextkRef2) {cand2 = cand; iFoundRefs++;}
+      if(tkRef == vertextkRef3) {cand3 = cand; iFoundRefs++;}
+    }
+	  
+    if(iFoundRefs < 3){
+      edm::LogError("HLTDisplacedtktktkFilter") << "HLTDisplacedtktktkFilter: ERROR: the tracks matched with the tktktk vertex tracks should be at least three by definition."  << std::endl;
+      return false;
+    } 	
+    // calculate three-track transverse momentum
+    math::XYZVector pperp(cand1->px() + cand2->px() + cand3->px(),
+                          cand1->py() + cand2->py() + cand3->py(),
+                          0.);
+
+    const reco::Vertex::Point& vpoint=displacedVertex.position();
+    //translate to global point, should be improved
+    GlobalPoint secondaryVertex (vpoint.x(), vpoint.y(), vpoint.z());
```

Comment:
there is no point in this transformation. Please remote it

Reply:
changed